### PR TITLE
Add compute_initial_revenue to cross-multiply revenue by country and sector

### PR DESCRIPTION
**Pull request by honj-uw**
- Add function `compute_initial_revenue` which cross-multiplies revenue values by country and sector to get `init_revenue`
- Rename "search_term" column to "iso3" and "utics" in all instances

### CODE REVIEW

File: backend/src/main.rs [hunk 1]
```diff
@@ -68,6 +70,61 @@ async fn load_uparams(filter: Vec<String>) -> DataFrame {
...
 
+// Computes the initial revenue by cross-multiplying country and sector revenue proportions
+async fn compute_initial_revenue(Json(inputs): Json<Value>) -> Result<DataFrame, PolarsError> {
+    // Convert JSON to Polars DataFrame
+    let df = JsonReader::new(Cursor::new(inputs.to_string()))
+        .with_json_format(JsonFormat::Json)
+        .finish()?;
+
+    // Convert DataFrame to LazyFrame to perform expression-based transformations
+    let lf = df.lazy();
```
Comment (natalie-unwritten) [inline backend/src/main.rs:81]: I feel like this should be outside the function, maybe even to become a function of its own if we constantly need to convert. Then this function should just be the polars steps on the lf.

File: backend/src/main.rs [hunk 2]
```diff
@@ -68,6 +70,61 @@ async fn load_uparams(filter: Vec<String>) -> DataFrame {
...
+    let country_data_with_constant = country_data.with_columns(vec![
+        lit(1).alias("join_key")
+    ]);
+
+    let sector_data_with_constant = sector_data.with_columns(vec![
+        lit(1).alias("join_key")
+    ]);
+
+// Perform cross join by joining on the constant column
+let cross_join_result = country_data_with_constant
```
Comment (natalie-unwritten) [inline backend/src/main.rs:100]: indentation

File: backend/src/main.rs [hunk 3]
```diff
@@ -68,6 +70,61 @@ async fn load_uparams(filter: Vec<String>) -> DataFrame {
...
+    .join(
+        sector_data_with_constant, 
+        vec![col("join_key")],     
+        vec![col("join_key")],     
+        JoinArgs::new(JoinType::Inner),
+    );
+
+    let result_df = cross_join_result
+    .select([
```
Comment (natalie-unwritten) [inline backend/src/main.rs:109]: should these be indented in 1 level more?

File: backend/src/main.rs [hunk 4]
```diff
@@ -68,6 +70,61 @@ async fn load_uparams(filter: Vec<String>) -> DataFrame {
...
+        // Rename "search_term" columms to iso3 and utics, respectively
+        col("search_term").alias("iso3"),
+        col("search_term_right").alias("utics"),
+        col("value"),
+        col("value_right"),
+        col("metric"),
+        col("metric_right"),
```
Comment (natalie-unwritten) [inline backend/src/main.rs:116]: if no columns are being dropped, is it possible to select all the columns, then add the renames and cross multiplication step afterwards?

File: backend/src/main.rs [hunk 5]
```diff
@@ -105,29 +162,49 @@ async fn compute(Json(inputs): Json<Value>) -> impl IntoResponse {
...
-        .into_iter()
-        // convert &str to String
-        .filter_map(|opt| opt.map(|s| s.to_string()))
-        .collect();
-
-    /* 3. Retrieve Uparams filtered by list from user data */
-    let df_uparams = load_uparams(filter).await;
-    print!("{:?}", df_uparams);
-    println!("{:?}", df);
+let iso3_series = df
```
Comment (natalie-unwritten) [inline backend/src/main.rs:165]: indentation here and below

File: backend/src/main.rs [hunk 6]
```diff
@@ -68,6 +70,61 @@ async fn load_uparams(filter: Vec<String>) -> DataFrame {
...
+        (col("value") / lit(100.0) * col("value_right") / lit(100.0) * total_revenue_expr)
+            .alias("init_revenue")
+    ])
+    .collect()
+    .map_err(|e| PolarsError::from(e))?;
+
+    println!("Column names: {:?}", result_df.get_column_names());
+    println!("{:?}", result_df);
+
+Ok(result_df) }
```
Comment (natalie-unwritten) [inline backend/src/main.rs:127]:
```suggestion
Ok(result_df) 
}
```
and indentation

File: backend/src/main.rs [hunk 7]
```diff
@@ -105,29 +162,49 @@ async fn compute(Json(inputs): Json<Value>) -> impl IntoResponse {
...
+.into_iter()
+.filter_map(|opt| opt.map(|s| s.to_string()))
+.collect();
+
+let utics_filter: Vec<String> = utics_series
+.str()
+.expect("Failed to access string data for utics")
+.into_iter()
+.filter_map(|opt| opt.map(|s| s.to_string()))
+.collect();
```
Comment (natalie-unwritten) [inline backend/src/main.rs:186]: I assume its not possible to do combine these? selecting both iso3 and utics columns in 1 series and then filtering on both cols?

If not could write as a function if its not too difficult

File: backend/src/main.rs [hunk 8]
```diff
@@ -105,29 +162,49 @@ async fn compute(Json(inputs): Json<Value>) -> impl IntoResponse {
...
+
+// Combine both filters for uparams
+let combined_filter: Vec<String> = iso3_filter.into_iter()
+.chain(utics_filter.into_iter())
+.collect();
```
Comment (natalie-unwritten) [inline backend/src/main.rs:191]: doesnt collect turn combined filter into a df?

File: backend/src/main.rs (hunk 4 above)
Comment (natalie-unwritten) [reply backend/src/main.rs:116]: Something like this:

```
df_renamed = df.select([
    pl.all(),
    pl.col("B").alias("New_B")  # Rename column 'B' to 'New_B'
])
```

File: backend/src/main.rs (hunk 8 above)
Comment (honj-uw) [reply backend/src/main.rs:191]: no, I think collect() gathers all  iterator elements into a new vector (Vec<String>)

so combined_filter will be a new vector concatenating iso3_filter and utics_filter

File: backend/src/main.rs (hunk 7 above)
Comment (honj-uw) [reply backend/src/main.rs:186]: added as function` create_string_filter`

File: backend/src/main.rs (hunk 4 above)
Comment (honj-uw) [reply backend/src/main.rs:116]: As just discussed - we want to select only these columns and theres a bunch that would need dropping if we select all. So have kept as is for now

File: backend/src/main.rs [hunk 9]
```diff
@@ -68,66 +80,127 @@ async fn load_uparams(filter: Vec<String>) -> DataFrame {
...
-    let df_uparams = load_uparams(filter).await;
-    print!("{:?}", df_uparams);
-    println!("{:?}", df);
+    /* 4. Retrieve Uparams filtered by list from user data */
+    let df_uparams = load_uparams(combined_filter).await;
+
+    /* 5. */
+
+    // Drop redundant value columns from initial df before joining
+    let df = df
```
Comment (natalie-unwritten) [inline backend/src/main.rs:193]: Arent these the columns that need to be multiplied together to get the projection?

**Pull request closed without merging**